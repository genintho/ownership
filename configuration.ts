import * as fs from "node:fs";
import yaml from 'js-yaml';

export interface Config {
  configuration: {
    path: string;
    stopFirstError: boolean;
    debug: boolean;
  };
  teams: {
    name: string;
  }[];
  features: {
    [key: string]: {
      description: string;
      owner: string;
      files: string[];
    };
  };
}

export function parseConfig(argv:{config:string, path:string, debug:boolean}): Config {
  // @ts-expect-error
  let configData: Config = {};
  try {
    const configFile = fs.readFileSync(argv.config, 'utf8');
    configData = yaml.load(configFile);

  } catch (e) {
    console.error("\nError loading or parsing config file:", argv.config);
    console.error(e.message);
    process.exit(1);
  }
  configData = configData || {};
  configData.configuration = configData.configuration || {};

  // DEBUG
  configData.configuration.debug = argv.debug || configData.configuration.debug || false;

  if (configData.configuration.debug) {
    console.debug("\nConfiguration data from", argv.config + ":");
    console.debug(configData);
  }

  let pathToAnalyze = argv.path || configData.configuration?.path || "./";
  if (pathToAnalyze.endsWith("/")) {
    pathToAnalyze = pathToAnalyze.slice(0, -1);
  }

  configData.configuration.path = pathToAnalyze;
  configData.configuration.stopFirstError = configData.configuration.stopFirstError || false;

  if (configData.configuration.debug) {
    console.debug("\nFinal Configuration:");
    console.debug(configData);
  }
  return configData;
}