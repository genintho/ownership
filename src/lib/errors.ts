export class OError {
	message(): string {
		throw new Error("Must be implemented");
	}
}

export class OErrorFileNoOwner extends OError {
	public readonly filePath: string;
	constructor(filePath: string) {
		super();
		this.filePath = filePath;
	}
	message(): string {
		return this.filePath + " has no owner";
	}
}

export class OErrorNothingToTest extends OError {
	message(): string {
		return "Nothing to test";
	}
}

export class OErrorDebugAndQuiet extends OError {
	message(): string {
		return "Option debug and quiet cannot be enabled at the same time";
	}
}

export class OErrorNoConfig extends OError {
	message() {
		return "No Configuration file found. Run the command `init` to generate one.";
	}
}

export class OErrorNoPaths extends OError {
	message() {
		return "A Path must be provided";
	}
}
