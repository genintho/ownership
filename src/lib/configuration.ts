import * as fs from "node:fs";
import { load as YamlLoad } from "js-yaml";
import { log } from "./log.ts";
import type { LogLevel } from "./log.ts";
import { OErrorNoPaths } from "./errors.ts";
import { minimatch } from "minimatch";

/**
 * Configuration object used by the tool.
 */
export type Configuration = {
	logLevel: LogLevel;
	pathToConfig: string;
	pathsToScan: string | string[];
	pathToBaseline: string;
	pathToExclude: RegExp[];
	teams: Teams;
	features: Features;
	version: number;
};

export type Features = {
	[key: string]: Feature;
};

export type Feature = {
	description?: string;
	files: string[];
	owner: string;
};

/**
 * Configuration option that can be stored in a file
 */
type StoredConfig = Partial<{
	version: number;
	logLevel: LogLevel;
	pathToBaseline: string;
	/**
	 * List of global patterns to exclude from the scan.
	 */
	exclude: string[];
	teams: Teams;
	features: Features;
}>;

export type ConfigurationOptions = {
	pathsToScan: string | string[];
	// optional
	version?: number;
	pathToConfig?: string;
	pathToBaseline?: string;
	logLevel?: LogLevel;
	// pathToConfigFile?: string;
	exclude?: string[];
};

export type Team = { name: string };
export type Teams = Team[];

export function combineConfig(input: StoredConfig & ConfigurationOptions): Configuration {
	return {
		version: input.version || 1,
		logLevel: input.logLevel || "info",
		pathsToScan: input.pathsToScan || ["."],
		pathToBaseline: input.pathToBaseline || ".ownership-todo.yml",
		pathToExclude: excludeToRegex(input.exclude || []),
		teams: input.teams || [],
		features: input.features || {},
		pathToConfig: input.pathToConfig || "ownership.yaml",
	};
}

export function config(options: ConfigurationOptions): Readonly<Configuration> {
	log.setLevel(options.logLevel);

	const fromFile = readConfigFile(options.pathToConfig);
	const c = combineConfig({ ...fromFile, ...options });

	log.setLevel(c.logLevel);

	if (c.pathsToScan.length === 0) {
		throw new OErrorNoPaths();
	}

	return c;
}

function readConfigFile(pathToConfigFile?: string): StoredConfig {
	if (!pathToConfigFile) {
		return {};
	}
	if (!fs.existsSync(pathToConfigFile)) {
		throw new Error("Configuration file not found: " + pathToConfigFile);
	}
	let configFileContent = null;
	try {
		configFileContent = fs.readFileSync(pathToConfigFile, "utf8");
	} catch (e: any) {
		throw new Error("Error reading configuration file: " + e.message);
	}

	if (configFileContent === null || configFileContent === undefined || configFileContent.trim() === "") {
		throw new Error("Configuration file is empty: " + pathToConfigFile);
	}

	let parsedConfig;
	try {
		parsedConfig = YamlLoad(configFileContent) as any;
	} catch (e: any) {
		throw new Error("Error parsing configuration file: " + e.message);
	}

	if (typeof parsedConfig !== "object" || parsedConfig === null) {
		throw new Error("Configuration file is not a valid object: " + pathToConfigFile);
	}

	const validatedConfig: StoredConfig = {};
	for (const key of Object.keys(parsedConfig)) {
		switch (key) {
			case "exclude":
			case "features":
			case "teams":
			case "version":
			case "logLevel":
				validatedConfig[key] = parsedConfig[key];
				break;
			default:
				log.warn(`Unknown configuration key: ${key}`);
				break;
		}
	}

	return validatedConfig;
}

export function initDefault(): StoredConfig {
	return {
		version: 1,
		logLevel: "info",
		pathToBaseline: ".ownership-todo.yml",
		exclude: [],
		teams: [],
		features: {},
	};
}

function excludeToRegex(excludes: string[] | null = []): RegExp[] {
	if (!excludes) {
		return [];
	}
	// @ts-expect-error
	return excludes.map((exclude) => {
		// @TODO detect if not array and log error is it is the case
		// eslint-disable-next-line no-control-regex
		const regex = /^(?!^(CON|PRN|AUX|NUL|COM[1-9]|LPT[1-9])(\..*)?$)[^<>:"/\\|?*\x00-\x1F]{1,255}[\\/]{0,1}$/i;
		if (regex.test(exclude)) {
			return new RegExp(exclude);
		}
		return minimatch.makeRe(exclude);
	});
}
