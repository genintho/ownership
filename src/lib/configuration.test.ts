import { describe, expect, it, beforeAll, afterAll } from "vitest";
import { config, initDefault } from "./configuration.ts";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import * as fs from "node:fs";
import { OErrorNoConfig, OErrorNoPaths } from "./errors.ts";
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

describe("parseConfig", () => {
	const testDir = path.join(__dirname, "test_configs");

	beforeAll(() => {
		if (!fs.existsSync(testDir)) {
			fs.mkdirSync(testDir, { recursive: true });
		}
	});

	afterAll(() => {
		if (fs.existsSync(testDir)) {
			fs.rmSync(testDir, { recursive: true, force: true });
		}
	});

	it("should throw an error if the config file does not exist", () => {
		expect(() =>
			config({
				pathToConfig: path.join(testDir, "nonexistent.yaml"),
				pathsToScan: ["./"],
			}),
		).toThrow(OErrorNoConfig);
	});

	it("should throw an error if the config file is invalid YAML", () => {
		const invalidYamlPath = path.join(testDir, "invalid.yaml");
		fs.writeFileSync(invalidYamlPath, "invalid yaml: content:"); // Malformed YAML

		expect(() => config({ pathToConfig: invalidYamlPath, pathsToScan: ["./"] })).toThrowError(
			/\nError loading or parsing config file: /,
		);
	});

	it("generate a default config", () => {
		expect(initDefault()).toMatchSnapshot();
	});

	it("should handle multiple paths", () => {
		const conf = config({ pathsToScan: ["./src", "./lib", "./test.js"] });
		expect(conf.pathsToScan).toHaveLength(3);
		expect(conf.pathsToScan[0]).toBe("src");
		expect(conf.pathsToScan[1]).toBe("lib");
		expect(conf.pathsToScan[2]).toBe("test.js");
	});

	it.only("a path must be set", () => {
		expect(() => config({ pathToConfig: "aa.yaml", pathsToScan: [] })).toThrow(OErrorNoPaths);
	});
});
