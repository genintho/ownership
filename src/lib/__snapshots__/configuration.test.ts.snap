// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseConfig > generate a default config 1`] = `
{
  "configuration": {
    "debug": true,
    "path": ".",
    "pathTodo": "./.owner-todo.yml",
    "stopFirstError": false,
    "updateTodo": false,
  },
  "exclude": [],
  "features": {},
  "teams": {},
}
`;
